File: 2022-09-30-react-env-variable-setting/backend/src/test/java/action/in/blog/todo/controller/TodoControllerTest.java
package action.in.blog.todo.controller;

import action.in.blog.todo.domain.Todo;
import action.in.blog.todo.service.TodoService;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.util.Arrays;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

class TodoControllerTest {

    TodoService mockTodoService;
    MockMvc sut;

    @BeforeEach
    void setUp() {
        mockTodoService = Mockito.mock(TodoService.class);
        sut = MockMvcBuilders.standaloneSetup(new TodoController(mockTodoService)).build();
    }

    @Test
    void receive_new_todo_and_delegate_command_to_service() throws Exception {

        ArgumentCaptor<Todo> argumentCaptor = ArgumentCaptor.forClass(Todo.class);
        ObjectMapper objectMapper = new ObjectMapper();
        Todo todo = new Todo();
        todo.setTitle("Hello World");


        sut.perform(
                post("/todo")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(todo))
        ).andExpect(status().isOk());


        verify(mockTodoService, times(1)).addTodo(argumentCaptor.capture());
        Todo capturedTodo = argumentCaptor.getValue();
        assertThat(capturedTodo.getTitle(), equalTo("Hello World"));
    }

    @Test
    void respond_todos() throws Exception {
        Todo todo = new Todo();
        todo.setId("1");
        todo.setTitle("Hello World");
        when(mockTodoService.getTodos()).thenReturn(Arrays.asList(todo));


        sut.perform(get("/todos"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$[0].id", equalTo("1")))
                .andExpect(jsonPath("$[0].title", equalTo("Hello World")))
        ;
    }
}

File: 2024-08-08-connect-aws-ecs-service-and-rds/action-in-blog/backend/src/test/kotlin/com/example/demo/controller/TodoControllerTest.kt
package com.example.demo.controller

import com.example.demo.domain.Todo
import com.example.demo.service.TodoService
import com.example.demo.util.TimeProvider
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito.*
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post
import org.springframework.test.web.servlet.setup.MockMvcBuilders

class TodoControllerTest {

    lateinit var mockTodoService: TodoService
    lateinit var mockMvc: MockMvc

    @BeforeEach
    fun setUp() {
        mockTodoService = mock(TodoService::class.java)
        mockMvc = MockMvcBuilders
            .standaloneSetup(
                TodoController(mockTodoService)
            )
            .build()
    }

    @Test
    fun `when request todo list then 200(status is OK)`() {
        mockMvc.get("/api/todos")
            .andExpect {
                status { isOk() }
            }
    }

    @Test
    fun `given data is existed when request todo list then return todo list`() {
        `when`(mockTodoService.getTodos())
            .thenReturn(
                listOf(
                    Todo(id = 1, title = "hello"),
                    Todo(id = 2, title = "world"),
                )
            )


        mockMvc.get("/api/todos")
            .andExpect {
                jsonPath("$[0].id") { value(1) }
                jsonPath("$[0].title") { value("hello") }
                jsonPath("$[1].id") { value(2) }
                jsonPath("$[1].title") { value("world") }
            }
        verify(mockTodoService, times(1))
            .getTodos()
    }

    @Test
    fun `when request todo creation then 200(status is OK)`() {
        val mockedTimeProvider = mockStatic(TimeProvider::class.java)
        mockedTimeProvider.`when`<Long> { TimeProvider.now() }.thenReturn(1722320449081)
        mockedTimeProvider.`when`<String> { TimeProvider.toStringFormat(1722320449081) }.thenReturn("2024-07-30 15:20:49")


        mockMvc.post("/api/todos") {
            contentType = MediaType.APPLICATION_JSON
            content = """
                {
                    "id": 0,
                    "title": "Hello World"
                }
            """.trimIndent()
        }.andExpect {
            status { isOk() }
        }
        verify(mockTodoService, times(1))
            .createTodo(
                Todo(id = 0, title = "Hello World", createdAt = "2024-07-30 15:20:49")
            )
        mockedTimeProvider.close()
    }
}
